packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts

```diff
@@ -208,9 +208,14 @@ export class PluginContributionHandler {
                             embeddedLanguages: this.convertEmbeddedLanguages(grammar.embeddedLanguages, logError),
                             tokenTypes: this.convertTokenTypes(grammar.tokenTypes)
                         }));
-                        pushContribution(`grammar.language.${language}.activation`,
-                            () => this.onDidActivateLanguage(language, () => this.monacoTextmateService.activateLanguage(language))
-                        );
+                    }
+                    if (grammars.indexOf(grammar) === (grammars.length - 1)) {
+                        if (languages && languages.length) {
+                            for (const lang of languages) {
+                                pushContribution(`language.${lang.id}.activation`,
+                                    () => this.onDidActivateLanguage(lang.id, () => this.monacoTextmateService.activateLanguage(lang.id)));
```

Comment:
I don't get how it helps, here we activate a language grammar when a language gets activated. It is no-op if there is no grammar.

Reply:
Is it about this listener callback: https://github.com/tom-shan/theia/blob/12085ee4076a4add4873d2a2d728ec6ef2fdc4a1/packages/plugin-ext/src/hosted/browser/hosted-plugin.ts#L162?

Reply:
For there is no listener of .ipynb files here and the onDidActivateLanguage() of .ipynb files will not be triggered.

Reply:
Yeah, got the issue, we are mixing language and grammar activation and notify a plugin system as a side effect of the grammar activation. We should decouple them, activated languages should be collected regardless of grammars at all. The proposed fix works only by luck because python extension has some grammar, if it won’t have it won’t work as well.

Reply:
I moved the entire setTimeout block out. Please review again.

Reply:
it has to be inside because of grammars. I will have a look later, i don't think it can be fixed by moving code in this place. Rather we should listen to language activation directly from monaco.